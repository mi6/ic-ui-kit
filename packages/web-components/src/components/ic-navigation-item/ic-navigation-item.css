@import "../../global/normalize.css";

/**
 * @prop --ic-z-index-navigation-item: z-index of navigation item
 */

:host {
  display: block;
}

:host(:focus-within) {
  z-index: var(--ic-z-index-navigation-item);
}

/* chevron */
svg {
  width: var(--ic-space-xl);
  fill: var(--ic-theme-text);
}

.svg {
  margin-left: auto;
}

:host(.expandable.navigation-item) .link,
:host(.expandable.navigation-item) ::slotted(a) {
  padding-right: 0;
}

:host(.navigation-item) .link,
:host(.navigation-item) ::slotted(a) {
  height: calc(100% - var(--ic-space-xxs));
  width: fit-content;
  color: var(--ic-theme-text);
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0 var(--ic-space-md) var(--ic-space-xxs);
  text-decoration: none;
  transition: var(--ic-easing-transition-fast);
  position: relative;
  white-space: nowrap;
}

:host(.navigation-item) ::slotted(a) {
  height: 100%;
  font: var(--ic-font-label) !important;
  letter-spacing: var(--ic-font-letter-spacing-0pt025) !important;
  padding: 0 var(--ic-space-md) !important;
}

:host(.navigation-item-selected) .link::after,
:host(.navigation-item-selected) ::slotted(a.active)::after,
:host(.navigation-item-top-nav) ::slotted(a.active)::after {
  content: "";
  height: 0.25rem;
  width: 100%;
  position: absolute;
  bottom: 0;
  background-color: var(--ic-theme-text);
}

:host(.navigation-item-selected.navigation-item-page-header)
  ::slotted(a.active)::after,
:host(.navigation-item-selected.navigation-item-page-header) .link::after,
:host(.navigation-item.navigation-item-page-header) ::slotted(a.active)::after,
:host(.navigation-item-top-nav-child) ::slotted(a.active)::after {
  content: none;
}

:host(.navigation-item-side-menu-selected) .link,
:host(.navigation-item-side-menu) ::slotted(a.active),
:host(.navigation-item-top-nav-child) ::slotted(a.active) {
  box-shadow: none;
}

:host(.navigation-item) .link:hover:not(:focus) {
  background-color: var(--ic-theme-hover);
  cursor: pointer;
}

:host(.navigation-item:not(.navigation-item-page-header, .navigation-item-side-menu))
  ::slotted(a:hover:not(:focus)) {
  background-color: var(--ic-theme-hover) !important;
}

:host(.navigation-item:not(.navigation-item-top-nav-child, .navigation-item-side-menu, .navigation-item-top-nav-child-selected
      .navigation-item-side-menu-selected))
  .focus-indicator:focus-within {
  box-shadow: var(--ic-border-focus);
  border-radius: var(--ic-border-radius);
}

:host(.navigation-item) .link:focus,
:host(.navigation-item) ::slotted(a:focus) {
  box-shadow: var(--ic-border-focus);
  border-radius: var(--ic-border-radius);
  outline: var(--ic-hc-focus-outline);
}

:host(.navigation-item-selected) .link:focus,
:host(.navigation-item) ::slotted(a.active:focus) {
  box-shadow: var(--ic-border-focus);
  border-radius: var(--ic-border-radius);
  outline: var(--ic-hc-focus-outline);
}

:host(.navigation-item-selected.dark) .link:focus,
:host(.navigation-item.dark) ::slotted(a.active:focus) {
  box-shadow: var(--ic-border-focus);
  border-radius: var(--ic-border-radius);
  outline: var(--ic-hc-focus-outline);
}

:host(.navigation-item) .link:active:not(:focus),
:host(.navigation-item) ::slotted(a:active:not(:focus)) {
  background-color: var(--ic-theme-active);
}

:host(.navigation-item-side-menu) .link,
:host(.navigation-item-side-menu) ::slotted(a) {
  height: 2.5rem;
  width: 100%;
  box-sizing: border-box;
  color: var(--ic-color-primary-text);
  display: flex;
  align-items: center;
  justify-content: flex-start;
  text-decoration: none;
  transition: box-shadow var(--ic-transition-duration-fast),
    border-radius var(--ic-transition-duration-fast);
}

:host(.navigation-item-side-menu) ::slotted(a) {
  font: var(--ic-font-body) !important;
  letter-spacing: var(--ic-font-letter-spacing-0pt005) !important;
  transition: box-shadow var(--ic-easing-transition-fast),
    background-color var(--ic-easing-transition-fast);
}

:host(.navigation-item-side-menu) .link:hover:not(:focus) {
  background-color: var(--ic-action-dark-bg-hover);
}

:host(.navigation-item-side-menu) ::slotted(a:hover:not(:focus)) {
  background-color: var(--ic-action-dark-bg-hover) !important;
}

:host(.navigation-item-side-menu) .focus-indicator:focus-within {
  box-shadow: none;
}

:host(.navigation-item-side-menu) .link:focus,
:host(.navigation-item-side-menu) ::slotted(a:focus) {
  box-shadow: var(--ic-border-focus-inset) !important;
  padding-right: var(--ic-space-lg);
  width: auto;
  border-radius: var(--ic-border-radius-inset) !important;
}

:host(.navigation-item-side-menu-selected) .link,
:host(.navigation-item-side-menu-selected) ::slotted(a) {
  transition: box-shadow var(--ic-easing-transition-slow),
    background-color var(--ic-easing-transition-slow);
}

:host(.navigation-item-side-menu-selected) .link:focus,
:host(.navigation-item-side-menu) ::slotted(a.active:focus),
:host(.navigation-item-top-nav-child) ::slotted(a.active:focus) {
  margin-left: var(--ic-space-xs) !important;
  padding-left: var(--ic-space-xs) !important;
}

:host(.navigation-item-top-nav-child) .link,
:host(.navigation-item-top-nav-child) ::slotted(a) {
  height: 2.5rem;
  width: fit-content;
  min-width: 9.063rem;
  color: var(--ic-color-primary-text);
  display: flex;
  align-items: center;
  justify-content: flex-start;
  padding: 0 var(--ic-space-md) !important;
  text-decoration: none;
  position: relative;
  transition: box-shadow var(--ic-transition-duration-fast),
    border-radius var(--ic-transition-duration-slow);
}

:host(.navigation-item-top-nav-child) ::slotted(a) {
  font: var(--ic-font-body) !important;
  letter-spacing: var(--ic-font-letter-spacing-0pt005) !important;
}

:host(.navigation-item-top-nav-child) .link:hover:not(:focus) {
  background-color: var(--ic-action-dark-bg-hover) !important;
}

:host(.navigation-item-top-nav-child) ::slotted(a:hover):not(:focus) {
  background-color: var(--ic-action-dark-bg-hover) !important;
}

:host(.navigation-item-top-nav-child) .link:focus,
:host(.navigation-item-top-nav-child) ::slotted(a:focus) {
  z-index: 1;
  outline: var(--ic-hc-focus-outline);
  border-radius: var(--ic-border-radius-inset);
  box-shadow: var(--ic-border-focus-inset);
  transition: border-radius 0s, box-shadow var(--ic-transition-duration-fast);
}

:host(.navigation-item-top-nav-child) ::slotted(a.active:focus) {
  box-shadow: var(--ic-border-focus-inset);
  border-radius: var(--ic-border-radius-inset);
}

:host(.navigation-item-side-menu) .link:active:not(:focus) {
  background-color: var(--ic-action-dark-bg-active);
}

:host(.navigation-item-top-nav-child) .link:active:not(:focus),
:host(.navigation-item-top-nav-child) ::slotted(a:active):not(:focus) {
  outline: var(--ic-hc-focus-outline);
  background-color: var(--ic-action-dark-bg-active);
}

:host(.navigation-item-top-nav-child) ::slotted(a:active):not(:focus),
:host(.navigation-item-side-menu) ::slotted(a:active):not(:focus) {
  background-color: var(--ic-action-dark-bg-active) !important;
}

:host(.navigation-item-side-menu-selected) *,
:host(.navigation-item-top-nav-child-selected) *,
:host(.navigation-item-side-menu) ::slotted(a.active),
:host(.navigation-item-top-nav-child) ::slotted(a.active) {
  font-weight: 700 !important;
}

:host(.navigation-item-top-nav-child) ::slotted(a.active)::before,
:host(.navigation-item-top-nav-child-selected) .link::before,
:host(.navigation-item-side-menu-selected) .link::before,
:host(.navigation-item-side-menu) ::slotted(a.active)::before,
:host(.navigation-item-side-menu-selected) .link:focus::before,
:host(.navigation-item-side-menu) ::slotted(a.active:focus)::before {
  content: "";
  position: absolute;
  left: 0;
  width: var(--ic-space-xs);
  height: 2.5rem;
  margin-top: 0.25rem;
  margin-right: 0.313rem;
  background-color: var(--ic-action-default);
  transition: left 0s;
}

:host(.navigation-item-top-nav-child-selected) .link:focus {
  margin-left: var(--ic-space-xs);
  margin-right: calc(-1 * var(--ic-space-xs));
  padding-left: var(--ic-space-xs) !important;
}

:host(.navigation-item-top-nav-child-selected) .link:focus::before {
  left: calc(-1 * var(--ic-space-xs));
  z-index: 1;
}

:host(.navigation-item-side-menu-selected) .link::before,
:host(.navigation-item-side-menu) ::slotted(a.active)::before,
:host(.navigation-item-side-menu-selected) .link:focus::before,
:host(.navigation-item-side-menu) ::slotted(a.active:focus)::before {
  z-index: 1;
}

:host(.navigation-item-side-menu-selected) .link:focus::before,
:host(.navigation-item-side-menu) ::slotted(a.active:focus)::before,
:host(.navigation-item-top-nav-child) ::slotted(a.active:focus)::before {
  left: calc(-1 * var(--ic-space-xs));
}

:host(.navigation-item-side-menu-selected) .link::after {
  display: none;
}

:host(.navigation-item-side-nav-collapsed) .link {
  height: 100%;
}

:host(.navigation-item-side-nav-collapsed) ::slotted(svg),
:host(.navigation-item-side-nav-collapsed-with-label) ::slotted(svg) {
  margin: auto;
  pointer-events: none;
}

:host(.navigation-item-side-nav-collapsed) .icon,
:host(.navigation-item-side-nav-collapsed-with-label) .icon,
:host(.navigation-item-side-nav-collapsed-with-label) ic-typography {
  align-items: center;
  padding: 0;
  margin: auto 0;
  width: fit-content;
  pointer-events: none;
}

:host(.navigation-item-side-nav-collapsed-with-label) ic-typography {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  width: 100%;
  text-align: center;
}

:host(.navigation-item-side-nav-collapsed-with-label) .link,
:host(.navigation-item-side-nav-collapsed-with-label) ::slotted(a) {
  display: flex;
  flex-direction: column;
  height: fit-content;
  padding: var(--ic-space-xs) !important;
  gap: var(--ic-space-xxs);
}

:host(.navigation-item-side-nav.navigation-item-side-nav-collapsed-with-label)
  .link,
:host(.navigation-item-side-nav.navigation-item-side-nav-collapsed-with-label)
  ::slotted(a) {
  height: auto !important;
  gap: 0;
}

:host(.navigation-item-side-nav.navigation-item-side-nav-collapsed-with-label)
  ::slotted(a) {
  display: block;
  text-align: center;
}

:host(.navigation-item-side-nav.navigation-item-side-nav-collapsed-with-label)
  .link
  ic-typography {
  opacity: 1;
}

:host(.navigation-item-side-nav-collapsed) svg,
:host(.navigation-item-side-nav-collapsed-with-label) svg {
  display: none;
}

::slotted(svg) {
  fill: var(--ic-theme-text);
  width: var(--ic-space-lg);
  margin-left: var(--ic-space-md);
}

.icon {
  height: var(--ic-space-lg);
  width: var(--ic-space-xxl);
  margin-left: calc(-1 * var(--ic-space-md));
  flex: 0 1 0%;
}

.icon > ::slotted(*) {
  margin-left: var(--ic-space-md) !important;
}

:host(.navigation-item-side-nav-collapsed) .icon > ::slotted(*) {
  margin: auto !important;
}

:host(.navigation-item-side-nav) .link,
:host(.navigation-item-side-nav) ::slotted(a) {
  height: var(--navigation-item-height);
  width: var(--navigation-item-width);
  justify-content: var(--navigation-item-justify-content);
  display: flex;
  gap: var(--ic-space-xs);
  box-sizing: border-box;
  color: var(--navigation-item-child-color) !important;
  text-decoration: none !important;
  white-space: nowrap;
  padding: 0 var(--ic-space-md);
}

:host(.navigation-item-side-nav.navigation-item-top-nav-child) .link,
:host(.navigation-item-side-nav.navigation-item-top-nav-child) ::slotted(a) {
  height: var(--navigation-item-child-height);
  width: var(--navigation-group-width);
  color: var(--navigation-item-child-color) !important;
  display: flex;
  gap: 0.625rem;
  box-sizing: border-box;
  min-width: 0;
  transition: box-shadow var(--ic-easing-transition-fast);
  text-decoration: none !important;
  white-space: nowrap;
}

:host(.navigation-item-side-nav.navigation-item-top-nav-child) .link:focus,
:host(.navigation-item-side-nav.navigation-item-top-nav-child)
  ::slotted(a:focus) {
  box-shadow: var(--ic-border-focus-dark);
}

:host(.navigation-item-side-nav.navigation-item-side-menu) .link:active {
  background-color: var(--navigation-item-child-active);
}

:host(.navigation-item-side-nav.navigation-item-top-nav-child) .link:active,
:host(.navigation-item-side-nav.navigation-item-top-nav-child)
  ::slotted(a:active) {
  background-color: var(--navigation-item-child-active);
}

:host(.navigation-item-side-nav) a.link ic-typography,
:host(.navigation-item-side-nav.navigation-item-top-nav-child)
  a.link
  ic-typography,
:host(.navigation-item-side-nav) div.link ic-typography,
:host(.navigation-item-side-nav.navigation-item-top-nav-child)
  div.link
  ic-typography,
.navigation-item-side-nav-slotted-text {
  opacity: var(--navigation-item-label-opacity);
  transition: opacity var(--ic-easing-transition-slow);
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

:host(.navigation-item-side-nav.navigation-item-selected) .link,
:host(.navigation-item-side-nav.navigation-item) ::slotted(a.active) {
  box-shadow: inset 0.313rem 0 0 var(--ic-theme-text);
}

:host(.navigation-item-side-nav.navigation-item-selected) .link::before,
:host(.navigation-item-side-nav) ::slotted(a.active)::before {
  content: "";
  display: block;
  position: absolute;
  inset: 0 0 0 0.313rem;
  border-radius: 0.688rem !important;
  transition: var(--ic-easing-transition-slow);
}

:host(.navigation-item-side-nav.navigation-item-selected) .link:focus::before,
:host(.navigation-item-side-nav) ::slotted(a.active:focus)::before {
  box-shadow: var(--ic-border-focus-inset);
}

:host(:not(.navigation-item-side-nav-collapsed).navigation-item-side-nav.navigation-item-side-nav-collapsed-with-label)
  .link {
  opacity: var(--navigation-item-label-opacity);
  height: auto;
}

:host(.navigation-item-top-nav-child.navigation-item-side-nav-collapsed-with-label)
  .link,
:host(.navigation-item-top-nav-child.navigation-item-side-nav-collapsed-with-label)
  ::slotted(a) {
  min-width: var(--navigation-group-item-min-width);
  padding: var(--ic-space-xs) !important;
  height: auto;
  gap: 0;
}

:host(.navigation-item-top-nav-child.navigation-item-side-nav-collapsed-with-label)
  ::slotted(a) {
  display: block;
}

:host(.navigation-item-side-nav.navigation-item-selected) .link::after,
:host(.navigation-item-side-nav) ::slotted(a.active)::after {
  display: none;
}

:host(.navigation-item-side-nav.navigation-item) .link:focus,
:host(.navigation-item-side-nav.navigation-item) ::slotted(a:focus) {
  box-shadow: var(--ic-border-focus-inset);
  border-radius: var(--ic-border-radius-inset);
  background-color: transparent;
}

:host(.navigation-item-side-nav.navigation-item-selected) .link:focus,
:host(.navigation-item-side-nav) ::slotted(a.active:focus) {
  margin: 0 auto;
  box-shadow: inset 0.313rem 0 0 var(--ic-theme-text);
  border-radius: 0;
}

:host(.navigation-item-side-nav.navigation-item) .link:hover:not(:focus) {
  background-color: var(--ic-theme-hover) !important;
}

:host(.navigation-item-page-header).link,
:host(.navigation-item-page-header) a,
:host(.navigation-item-page-header) ::slotted(a) {
  height: 2.5rem !important;
  color: var(--ic-color-primary-text) !important;
  transition: all var(--ic-easing-transition-fast) !important;
  box-shadow: rgba(23 89 188 / 0%) !important;
}

:host(.navigation-item-page-header.navigation-item-selected) .link::after,
:host(.navigation-item-page-header) ::slotted(a.active)::after {
  content: " " !important;
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  height: var(--ic-space-xxs);
  background-color: var(--ic-action-default) !important;
}

:host(.navigation-item-page-header) .link:hover,
:host(.navigation-item-page-header) ::slotted(.link:hover),
:host(.navigation-item-page-header) ::slotted(a:hover),
:host(.navigation-item-page-header) .link:hover:not(:focus) {
  background-color: var(--ic-action-default-bg-hover) !important;
  cursor: pointer;
}

:host(.navigation-item-page-header) .link:focus,
:host(.navigation-item-page-header) ::slotted(a:focus) {
  box-shadow: var(--ic-border-focus);
}

:host(.navigation-item-page-header) .link:active,
:host(.navigation-item-page-header) ::slotted(a:active) {
  background-color: var(--ic-action-default-bg-active) !important;
}

:host(.navigation-item-page-header.navigation-item-selected) .link:focus::after,
:host(.navigation-item-page-header) ::slotted(a.active:focus)::after {
  border-bottom-left-radius: var(--ic-border-radius);
  border-bottom-right-radius: var(--ic-border-radius);
}

:host(.navigation-item-page-header.navigation-item-selected.with-transition)
  .link,
:host(.navigation-item-page-header.navigation-item-selected.with-transition)
  ::slotted(a.active) {
  transition: all var(--ic-easing-transition-fast),
    border-color var(--ic-transition-duration-fast);
}

:host(.navigation-item-page-header) ::slotted(a.active)::before,
:host(.navigation-item-page-header) ::slotted(a.active:focus)::before,
:host(.navigation-item-page-header.navigation-item-selected)
  .link:focus::before {
  display: none;
}

.open-in-new-icon {
  flex: 1;
}

.open-in-new-icon svg {
  margin-top: var(--ic-space-xxs);
  padding-left: var(--ic-space-xxs);
}

:host(.navigation-item-side-nav) .open-in-new-icon {
  margin-left: calc(-1 * var(--ic-space-xs));
}

:host(.navigation-item-top-nav) .open-in-new-icon svg,
:host(.navigation-item-page-header) .open-in-new-icon svg,
:host(.navigation-item-side-nav) .open-in-new-icon svg {
  height: 0.875rem;
  width: 0.875rem;
}

:host(.navigation-item-side-menu) .open-in-new-icon svg,
:host(.navigation-item-top-nav-child) .open-in-new-icon svg {
  height: var(--ic-space-md);
  width: var(--ic-space-md);
}

:host(.navigation-item-page-header) .open-in-new-icon svg,
:host(.navigation-item-top-nav-child) .open-in-new-icon svg,
:host(.navigation-item-side-menu) .open-in-new-icon svg {
  fill: var(--ic-color-primary-text);
}
