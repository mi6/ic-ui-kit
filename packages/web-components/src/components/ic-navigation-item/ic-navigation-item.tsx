import {
  Component,
  Element,
  Event,
  EventEmitter,
  Listen,
  Method,
  h,
  Host,
  Prop,
  State,
} from "@stencil/core";

import {
  DEVICE_SIZES,
  getCurrentDeviceSize,
  getThemeForegroundColor,
  getParentElementType,
  getNavItemParentDetails,
} from "../../utils/helpers";
import { IcNavType, IcTheme } from "../../utils/types";

import chevronIcon from "../../assets/chevron-icon.svg";

import OpenInNew from "../../assets/OpenInNew.svg";

/**
 * @part link - The `<a>` within ic-navigation-item
 */

@Component({
  tag: "ic-navigation-item",
  styleUrl: "ic-navigation-item.css",
  shadow: {
    delegatesFocus: true,
  },
})
export class NavigationItem {
  private navigationSlot: HTMLElement;
  private isInitialRender: boolean = true;
  private itemEl: HTMLElement;

  @Element() el: HTMLIcNavigationItemElement;

  @State() deviceSize: number = DEVICE_SIZES.XL;
  @State() focusStyle = getThemeForegroundColor();
  @State() inTopNavSideMenu: boolean = false;
  @State() isSideNavMobile: boolean = false;
  @State() isTopNavChild: boolean = false;
  @State() navigationType: IcNavType | "";
  @State() parentEl: HTMLElement;
  @State() sideNavExpanded: boolean = false;

  /**
   * @internal If `true`, the icon and label will be displayed when side navigation is collapsed.
   */
  @Prop() collapsedIconLabel: boolean = false;

  /**
   * @internal If `true`, the navigation item will be displayed within a tooltip.
   */
  @Prop() displayNavigationTooltip: boolean = false;

  /**
   * If `true`, the user can save the linked URL instead of navigating to it.
   */
  @Prop() download?: string | boolean = false;

  /**
   *  @internal If `true`, the navigation item will be expandable.
   */
  @Prop() expandable: boolean = false;

  /**
   * The destination of the navigation item.
   */
  @Prop() href: string = "";

  /**
   * The human language of the linked URL.
   */
  @Prop() hreflang?: string;

  /**
   * The label of the navigation item.
   */
  @Prop() label: string;

  /**
   * The relationship of the linked URL as space-separated link types.
   */
  @Prop() rel?: string;

  /**
   * How much of the referrer to send when following the link.
   */
  @Prop() referrerpolicy?: ReferrerPolicy;

  /**
   *  If `true`, the navigation item will be set in a selected state.
   */
  @Prop() selected: boolean = false;

  /**
   *  The place to display the linked URL, as the name for a browsing context (a tab, window, or iframe).
   */
  @Prop() target?: string;

  /**
   * @internal - Emitted when item loses focus.
   */
  @Event() childBlur: EventEmitter<void>;

  /**
   * @internal - Emitted when navigation item clicked.
   */
  @Event() navItemClicked: EventEmitter<void>;

  disconnectedCallback(): void {
    if (this.navigationType === "side") {
      this.parentEl.removeEventListener(
        "sideNavExpanded",
        this.sideNavExpandHandler
      );
    } else if (this.navigationType === "top") {
      this.parentEl.removeEventListener(
        "topNavResized",
        this.topNavResizedHandler
      );
    }
  }

  componentWillLoad(): void {
    const navParentDetails = getNavItemParentDetails(this.el);
    this.navigationType = navParentDetails.navType;
    this.parentEl = navParentDetails.parent;

    if (this.navigationType === "side") {
      this.parentEl.addEventListener(
        "sideNavExpanded",
        this.sideNavExpandHandler
      );
    } else if (this.navigationType === "top") {
      this.parentEl.addEventListener(
        "topNavResized",
        this.topNavResizedHandler
      );
    }

    if (
      getParentElementType(this.el) === "IC-NAVIGATION-GROUP" &&
      this.navigationType === "top"
    ) {
      this.isTopNavChild = true;
    }

    this.deviceSize = getCurrentDeviceSize();
    if (this.deviceSize <= DEVICE_SIZES.L && this.navigationType === "top") {
      this.inTopNavSideMenu = true;
    }

    this.navigationSlot = this.el.querySelector('[slot="navigation-item"]');
    if (this.navigationSlot) {
      this.navigationSlot.ariaLabel = this.navigationSlot.textContent.trim();
    }
  }

  componentDidUpdate(): void {
    this.isInitialRender = false;
  }

  @Listen("themeChange", { target: "document" })
  themeChangeHandler(ev: CustomEvent): void {
    const theme: IcTheme = ev.detail;
    this.focusStyle = theme.mode;
  }

  /**
   * Sets focus on the nav item.
   */
  @Method()
  async setFocus(): Promise<void> {
    if (this.itemEl) {
      this.itemEl.focus();
    }
  }

  private displayDefaultNavigationItem = (
    href: string,
    hreflang: string,
    target: string,
    rel: string,
    referrerpolicy: ReferrerPolicy,
    download: string | boolean,
    label: string
  ): HTMLElement => {
    const variant =
      this.isTopNavChild || this.inTopNavSideMenu ? "body" : "label";
    const ChevronIconComponent = this.expandable && (
      <div class={{ svg: true }} innerHTML={chevronIcon}></div>
    );
    const IconComponent = this.el.querySelector('[slot="icon"]') && (
      <div class="icon">
        {" "}
        <slot name="icon"></slot>{" "}
      </div>
    );

    if (href !== "") {
      return (
        <a
          href={href}
          target={target}
          rel={rel}
          hreflang={hreflang}
          referrerPolicy={referrerpolicy}
          download={download !== false ? download : null}
          class="link"
          ref={(el) => (this.itemEl = el)}
          part="link"
        >
          {IconComponent}

          <ic-typography variant={variant}>{label}</ic-typography>
          {ChevronIconComponent}
          {target === "_blank" && (
            <span class="open-in-new-icon" innerHTML={OpenInNew} />
          )}
        </a>
      );
    }

    return (
      <div tabindex="0" class="link" ref={(el) => (this.itemEl = el)}>
        {IconComponent}
        <ic-typography variant={variant}>{label}</ic-typography>
        {ChevronIconComponent}
      </div>
    );
  };

  private topNavResizedHandler = (ev: CustomEvent): void => {
    const newSize = ev.detail.size;
    if (newSize !== this.deviceSize) {
      this.deviceSize = newSize;
      this.inTopNavSideMenu = newSize <= DEVICE_SIZES.L;
    }
  };

  private sideNavExpandHandler = (ev: CustomEvent): void => {
    const { sideNavExpanded, sideNavMobile } = ev.detail;
    this.sideNavExpanded = sideNavExpanded;
    this.isSideNavMobile = sideNavMobile;
  };

  private handleBlur = (ev: FocusEvent) => {
    if (ev.relatedTarget !== null) {
      const target = ev.relatedTarget as HTMLElement;
      if (target.tagName === "IC-NAVIGATION-ITEM") {
        return;
      }
    }
    this.childBlur.emit();
  };

  private handleClick = () => {
    this.navItemClicked.emit();
  };

  render() {
    const {
      href,
      hreflang,
      target,
      rel,
      referrerpolicy,
      download,
      label,
      inTopNavSideMenu,
      isTopNavChild,
      selected,
      navigationSlot,
    } = this;

    return (
      <Host
        class={{
          ["navigation-item"]: true,
          ["navigation-item-top-nav"]:
            !inTopNavSideMenu && this.navigationType === "top",
          ["navigation-item-top-nav-child-selected"]:
            isTopNavChild && !inTopNavSideMenu && selected,
          [this.focusStyle]:
            (!inTopNavSideMenu && !isTopNavChild) ||
            (this.navigationType === "side" && isTopNavChild),
          ["navigation-item-selected"]: !isTopNavChild && selected,
          ["navigation-item-side-menu"]: inTopNavSideMenu,
          ["navigation-item-side-menu-selected"]: inTopNavSideMenu && selected,
          ["navigation-item-top-nav-child"]: isTopNavChild && !inTopNavSideMenu,
          ["navigation-item-page-header"]:
            this.navigationType === "page-header",
          ["with-transition"]: !this.isInitialRender,
          ["navigation-item-side-nav"]: this.navigationType === "side",
          ["navigation-item-side-nav-collapsed"]:
            (!this.sideNavExpanded || this.displayNavigationTooltip) &&
            this.navigationType === "side",
          ["navigation-item-side-nav-collapsed-with-label"]:
            !this.sideNavExpanded &&
            this.navigationType === "side" &&
            this.collapsedIconLabel &&
            !this.isSideNavMobile,
          ["expandable"]: this.expandable,
        }}
        onBlur={isTopNavChild && !inTopNavSideMenu ? this.handleBlur : null}
        onClick={this.handleClick}
        aria-current={selected ? "page" : null}
        role="listitem"
      >
        {/* Tooltip enabled by applying navigation-item-side-nav-collapsed class to host */}
        <ic-tooltip
          label={label || navigationSlot.textContent}
          target="navigation-item"
          placement="right"
          class={{
            ["tooltip-navigation-item"]: true,
            ["tooltip-navigation-item-side-nav-collapsed"]:
              (!this.sideNavExpanded || this.displayNavigationTooltip) &&
              this.navigationType === "side",
            ["tooltip-long-label-navigation-item-side-nav-expanded"]:
              this.el.hasAttribute("[display-navigation-tooltip = 'true']"),
          }}
        >
          {navigationSlot ? (
            <slot name="navigation-item"></slot>
          ) : (
            this.displayDefaultNavigationItem(
              href,
              hreflang,
              target,
              rel,
              referrerpolicy,
              download,
              label
            )
          )}
        </ic-tooltip>
      </Host>
    );
  }
}
