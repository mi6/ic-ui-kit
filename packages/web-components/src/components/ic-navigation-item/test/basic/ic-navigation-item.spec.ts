import { newSpecPage } from "@stencil/core/testing";
import { DEVICE_SIZES } from "../../../../utils/helpers";
import { NavigationItem } from "../../ic-navigation-item";
import { TopNavigation } from "../../../ic-top-navigation/ic-top-navigation";
import { PageHeader } from "../../../ic-page-header/ic-page-header";
import * as helpers from "../../../../utils/helpers";

describe("ic-navigation-item", () => {
  it("renders with label", async () => {
    const page = await newSpecPage({
      components: [NavigationItem],
      html: `<ic-navigation-item label="Item label"></ic-navigation-item>`,
    });
    expect(page.root).toMatchSnapshot("renders-with-label");
  });

  it("should test theme change", async () => {
    const page = await newSpecPage({
      components: [NavigationItem],
      html: `<ic-navigation-item label="Item label"></ic-navigation-item>`,
    });

    await page.rootInstance.themeChangeHandler({ detail: { mode: "dark" } });
    await page.waitForChanges();
    expect(page.rootInstance.focusStyle).toBe("dark");

    await page.rootInstance.themeChangeHandler({ detail: { mode: "light" } });
    await page.waitForChanges();
    expect(page.rootInstance.focusStyle).toBe("light");
  });

  it("should test top nav resize handler", async () => {
    const page = await newSpecPage({
      components: [NavigationItem],
      html: `<ic-navigation-item label="Item label"></ic-navigation-item>`,
    });

    await page.rootInstance.topNavResizedHandler({
      detail: { size: DEVICE_SIZES.S },
    });
    await page.waitForChanges();
    expect(page.rootInstance.deviceSize).toBe(DEVICE_SIZES.S);
    expect(page.rootInstance.inTopNavSideMenu).toBe(true);
  });

  it("should test side nav expand handler", async () => {
    const page = await newSpecPage({
      components: [NavigationItem],
      html: `<ic-navigation-item label="Item label"></ic-navigation-item>`,
    });

    await page.rootInstance.setFocus();
    await page.rootInstance.sideNavExpandHandler({
      detail: { sideNavExpanded: true, sideNavMobile: true },
    });
    await page.waitForChanges();
    expect(page.rootInstance.sideNavExpanded).toBe(true);
    expect(page.rootInstance.isSideNavMobile).toBe(true);
  });

  it("should test private methods", async () => {
    const page = await newSpecPage({
      components: [NavigationItem],
      html: `<ic-navigation-item label="Item label"></ic-navigation-item>`,
    });

    page.root.click();
  });

  it("test inside side navigation", async () => {
    const page = await newSpecPage({
      components: [TopNavigation, NavigationItem],
      html: `<ic-side-navigation version="v0.0.0" app-title="ACME" status="BETA" collapsed-icon-labels="true">
      <ic-navigation-item slot="primary-navigation" href="/" label="Home">
      </ic-navigation-item>
    </ic-side-navigation>`,
    });
    await page.waitForChanges();
    await new Promise((r) => setTimeout(r, 2000));

    // test disconnected callback
    page.setContent("");
  });

  it("should render inside page header", async () => {
    const page = await newSpecPage({
      components: [PageHeader, NavigationItem],
      html: `
      <ic-page-header heading="Coffee recipes" subheading="This is a simple page header component and this is the text.">
        <ic-status-tag slot="heading-adornment" label="Beta"></ic-status-tag>
        <ic-navigation-item
          slot="tabs"
          label="All recipes"
          href="/all-recipes"
          selected
        ></ic-navigation-item>
        <ic-navigation-item
          slot="tabs"
          label="Favourites"
          href="/favourites"
        ></ic-navigation-item>
      </ic-page-header>`,
    });

    await page.waitForChanges();
    await new Promise((r) => setTimeout(r, 2000));

    // test disconnected callback
    page.setContent("");
  });

  // NOTE: This must go last as mocks getCurrentDeviceSize function, which will apply to all subsequent tests
  it("test inside top navigation", async () => {
    const myfunc = jest.fn().mockReturnValue(DEVICE_SIZES.S);
    Object.defineProperty(helpers, "getCurrentDeviceSize", {
      value: myfunc,
    });

    const page = await newSpecPage({
      components: [TopNavigation, NavigationItem],
      html: `<ic-top-navigation
        app-title="ApplicationName"
        status="alpha"
        version="v0.0.7"
      >
        <ic-navigation-item slot="navigation" label="Navigation 1" href="/"></ic-navigation-item>
      </ic-top-navigation>`,
    });
    await page.waitForChanges();
    await new Promise((r) => setTimeout(r, 2000));

    // test disconnected callback
    page.setContent("");
  });
});
