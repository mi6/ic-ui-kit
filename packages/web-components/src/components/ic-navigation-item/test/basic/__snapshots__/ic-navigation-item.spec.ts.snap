// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-navigation-item renders with label: renders-with-label 1`] = `
<ic-navigation-item class="light navigation-item" label="Item label" role="listitem">
  <mock:shadow-root>
    <ic-tooltip class="tooltip-navigation-item" label="Item label" placement="right" target="navigation-item">
      <div aria-labelledby="ic-tooltip-navigation-item" class="link" tabindex="0">
        <ic-typography variant="label">
          Item label
        </ic-typography>
      </div>
    </ic-tooltip>
  </mock:shadow-root>
</ic-navigation-item>
`;
